Userland/Libraries/LibCore/FileWatcher.h

```diff
@@ -13,48 +13,114 @@
 #include <AK/RefCounted.h>
 #include <AK/Result.h>
 #include <AK/String.h>
+#include <Kernel/API/InodeWatcherEvent.h>
+#include <Kernel/API/InodeWatcherFlags.h>
 #include <LibCore/Notifier.h>
 
 namespace Core {
 
 struct FileWatcherEvent {
     enum class Type {
-        Modified,
-        ChildAdded,
-        ChildRemoved,
+        MetadataModified,
+        ContentModified,
+        Deleted,
+        ChildCreated,
+        ChildDeleted,
     };
     Type type;
-    String child_path;
+    String event_path;
 };
 
-class BlockingFileWatcher {
+class FileWatcherBase {
+public:
+    virtual ~FileWatcherBase() { }
+
+    virtual String class_name() const = 0;
+
+    Result<bool, String> add_watch(String path, InodeWatcherEvent::Type event_mask);
+    Result<bool, String> remove_watch(String path);
+    bool is_watching(String path) { return m_path_to_wd.find(path) != m_path_to_wd.end(); }
+
+protected:
+    FileWatcherBase(int watcher_fd)
+        : m_watcher_fd(watcher_fd)
+    {
+    }
+
+    int m_watcher_fd { -1 };
+    HashMap<String, unsigned> m_path_to_wd;
+    HashMap<unsigned, String> m_wd_to_path;
+};
+
+class BlockingFileWatcher final : public FileWatcherBase {
     AK_MAKE_NONCOPYABLE(BlockingFileWatcher);
 
 public:
-    explicit BlockingFileWatcher(const String& path);
+    explicit BlockingFileWatcher(InodeWatcherFlags = InodeWatcherFlags::None);
     ~BlockingFileWatcher();
+    virtual String class_name() const override { return "BlockingFileWatcher"; }
 
     Optional<FileWatcherEvent> wait_for_event();
-
-private:
-    String m_path;
-    int m_watcher_fd { -1 };
 };
 
-class FileWatcher : public RefCounted<FileWatcher> {
+class FileWatcher final : public FileWatcherBase
+    , public RefCounted<FileWatcher> {
     AK_MAKE_NONCOPYABLE(FileWatcher);
 
 public:
-    static Result<NonnullRefPtr<FileWatcher>, String> watch(const String& path);
+    static Result<NonnullRefPtr<FileWatcher>, String> create(InodeWatcherFlags = InodeWatcherFlags::None);
     ~FileWatcher();
 
+    virtual String class_name() const override { return "FileWatcher"; }
+
     Function<void(FileWatcherEvent)> on_change;
```

Comment:
Now that `FileWatcherEvent` has a `String` member, we should change this signature to avoid copy construction:
```suggestion
    Function<void(FileWatcherEvent const&)> on_change;
```